axlearn/common/mixture_of_experts_neuron_test.py

```diff
@@ -31,768 +32,1105 @@
+        
+        if batch % out_batch != 0:
+            continue
+
+        config = builder.reset().with_dimensions(batch, seq, input_dim).with_expert_settings(
+                    hidden_dim,
+                    out_batch,
+                    n_groups,
+                    n_experts,
+                    capacity,
+                    train_capacity_factor=None
+                ).build_test_configs_moe()
+        name = f"b{batch}_s{seq}_i{input_dim}_h{hidden_dim}_e{n_experts}_g{n_groups}_ob{out_batch}_ec{capacity}"
+        test_configs.extend([(name + cfg.prefix, cfg) for cfg in config])
+
+    return test_configs
 
-        layer: Top2Gating = cfg.instantiate(parent=None)
-        state = layer.initialize_parameters_recursively(prng_key=jax.random.PRNGKey(123))
-        logits = jax.random.uniform(jax.random.PRNGKey(1), shape=(2, 4, group_size, num_experts))
+# pylint: disable=no-self-use,protected-access
+class TestImplCorrectness(TestCase):
 
-        @jax.jit
-        def top2_gating_with_adaptive_load_balance_loss():
-            return F(
+    def _fwd_call(self, layer, state, inputs):
+        return F(
                 layer,
                 is_training=True,
                 prng_key=jax.random.PRNGKey(123),
                 state=state,
-                inputs=dict(logits=logits),
-            )
-        gating, output_collection = top2_gating_with_adaptive_load_balance_loss()
-
-        self.assertIn("adaptive_load_balance_loss", output_collection.summaries)
-        self.assertContainsSubset(
-            {"scale", "value_average"}, output_collection.summaries["adaptive_load_balance_loss"]
-        )
-        # We return the adjusted load balance loss.
-        self.assertEqual(
-            output_collection.summaries["load_balance_loss"],
-            gating.load_balance_loss,
-        )
-        # The adjusted load balance loss is greater than the original load balance loss.
-        self.assertGreater(
-            output_collection.summaries["load_balance_loss"],
-            output_collection.summaries["load_balance_loss_original"],
+                inputs=inputs,
         )
 
-    @parameterized.parameters(
-        "prenorm",
-        "postnorm",
-        "hybridnorm",
-        "nonorm",
-    )
-    def test_layer_structure(self, structure):
-        batch, seq_len, dim = 2, 3, 4
-        cfg = TransformerFeedForwardMoE.default_config().set(
-            name="moe",
-            input_dim=dim,
-            hidden_dim=dim * 4,
-            activation="nn.relu",
-            structure=structure,
-            num_experts=4,
-            num_groups=2,
-        )
-        layer = cfg.instantiate(parent=None)
-        layer_params = layer.initialize_parameters_recursively(prng_key=jax.random.PRNGKey(0))
-        inputs = jax.random.normal(jax.random.PRNGKey(1), shape=[batch, seq_len, dim])
+    @parameterized.named_parameters(_get_training_configs())
```

Comment:
This redirection seems unnecessarily complicated. Doesn't it make it harder to add my own test for a single config? Also could you add an example for such a single config test.